grpc/server.go

```diff
@@ -0,0 +1,251 @@
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (response interface{}, err error) {
+	d.Logger.Debug("comparing with %s with %d endpoints", info.FullMethod, len(d.Configuration))
+
+	// FullMethod is the full RPC method string, i.e., /package.service/method.
+	if endptConfig, ok := d.Configuration[TargetEndpoint(info.FullMethod)]; ok {
+		randomPercent := rand.Intn(100)
+
+		if len(endptConfig.AlterationHash) > randomPercent {
+			altConfig := endptConfig.AlterationHash[randomPercent]
+
+			if altConfig.ErrorToReturn != "" {
+				d.Logger.Debug("Error Code: %s", v1beta1.ErrorMap[altConfig.ErrorToReturn])
+
+				return nil, status.Error(
+					v1beta1.ErrorMap[altConfig.ErrorToReturn],
+					// Future Work: interview users about this message //nolint:golint
+					fmt.Sprintf("Chaos-Controller injected this error: %s", altConfig.ErrorToReturn),
+				)
+			} else if altConfig.OverrideToReturn != "" {
+				d.Logger.Debug("OverrideToReturn: %s", altConfig.OverrideToReturn)
+
+				return &emptypb.Empty{}, nil
+			}
+
+			d.Logger.Error("Endpoint %s should define either an ErrorToReturn or OverrideToReturn but does not", endptConfig.TargetEndpoint)
+		}
+	}
+
+	return handler(ctx, req)
+}
+
+// GetAlterationToPercentAffected takes a series of alterations configured for a Spec and maps them to the percentage of queries which should be affected
+func GetAlterationToPercentAffected(endpointSpecList []*pb.AlterationSpec) (map[AlterationConfiguration]PercentAffected, error) {
+	// object returned indicates, for a particular AlterationConfiguration, what percentage of queries to which it should apply
+	mapping := make(map[AlterationConfiguration]PercentAffected)
+
+	// unquantified is a holding variable used later to calculate and assign percentages to alterations which do not specify queryPercent
+	unquantifiedAlts := make([]AlterationConfiguration, 0)
+
+	pctClaimed := 0
+
+	for _, altSpec := range endpointSpecList {
+		if altSpec.ErrorToReturn == "" && altSpec.OverrideToReturn == "" {
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying either ErrorToReturn or OverrideToReturn for all target endpoints")
+		}
+
+		if altSpec.ErrorToReturn != "" && altSpec.OverrideToReturn != "" {
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption where ErrorToReturn or OverrideToReturn are both specified for a target endpoints")
+		}
+
+		alterationConfig := AlterationConfiguration{
+			ErrorToReturn:    altSpec.ErrorToReturn,
+			OverrideToReturn: altSpec.OverrideToReturn,
+		}
+
+		// Intuition:
+		// (1) add all endpoints where queryPercent is specified
+		// (2) track percentage of queries already claimed by an endpointConfiguration
+		// (3) track endpoints where queryPercent is missing and calculate them later
+		if altSpec.QueryPercent > 0 {
+			mapping[alterationConfig] = PercentAffected(altSpec.QueryPercent)
+			pctClaimed += int(altSpec.QueryPercent)
```

Comment:
It would be, I think, very easy to check if we exceed 100 here and return an error. Why choose not to do so? That would help make some of the unit tests a little more clear, specifically the ones that check the certain error cases which we choose not to catch.